### PR TITLE
refactor: complete dtype rules for expression tree transformations

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Review by chelsea-lin — APPROVED** [review]
After learning more details offline, overall LGTM!

File: third_party/bigframes_vendored/pandas/core/generic.py [hunk 1]
```diff
@@ -756,12 +756,12 @@ def isna(self) -> NDFrame:
...
             >>> ser
-            0     5.0
+            0       5
             1    <NA>
-            2     6.0
+            2       6
             3    <NA>
             4    <NA>
-            dtype: Float64
+            dtype: Int64
```
Comment (chelsea-lin) [inline third_party/bigframes_vendored/pandas/core/generic.py:764]: Good to see that this change fix this issue!

File: bigframes/session/__init__.py [hunk 2]
```diff
@@ -923,25 +939,31 @@ def _read_pandas(
...
-        return dataframe.DataFrame(blocks.Block.from_local(pandas_dataframe, self))
+    ) -> Optional[dataframe.DataFrame]:
+        if pandas_dataframe.size > MAX_INLINE_DF_SIZE:
+            return None
+
+        try:
+            inline_df = dataframe.DataFrame(
+                blocks.Block.from_local(pandas_dataframe, self)
+            )
+        except ValueError:  # Thrown by ibis for some unhandled tyeps
```
Comment (chelsea-lin) [inline bigframes/session/__init__.py:957]: nit: types.

File: bigframes/operations/aggregations.py [hunk 3]
```diff
@@ -146,6 +205,24 @@ def skips_nulls(self):
     def handles_ties(self):
         return True
 
+    def output_type(self, *input_types: dtypes.ExpressionType):
+        if isinstance(self.bins, int) and (self.labels is False):
+            return dtypes.INT_DTYPE
```
Comment (chelsea-lin) [inline bigframes/operations/aggregations.py:210]: do you have a chance to test it? Is it `FLOAT_DTYPE`?

File: bigframes/operations/aggregations.py [hunk 4]
```diff
@@ -87,16 +112,38 @@ class ApproxQuartilesOp(UnaryAggregateOp):
...
+        else:
+            return input_types[0]
+
 
 @dataclasses.dataclass(frozen=True)
 class MeanOp(UnaryAggregateOp):
     name: ClassVar[str] = "mean"
 
+    def output_type(self, *input_types: dtypes.ExpressionType):
+        if pd.api.types.is_bool_dtype(input_types[0]):
```
Comment (chelsea-lin) [inline bigframes/operations/aggregations.py:129]: nit: 
```
if pd.api.types.is_bool_dtype(input_types[0]) or pd.api.types.is_integer_dtype(input_types[0]):
   return dtypes.FLOAT_DTYPE
```

File: bigframes/operations/aggregations.py [hunk 5]
```diff
@@ -87,16 +112,38 @@ class ApproxQuartilesOp(UnaryAggregateOp):
     def name(self):
         return f"{self.quartile*25}%"
 
+    def output_type(self, *input_types: dtypes.ExpressionType):
+        if pd.api.types.is_bool_dtype(input_types[0]):
```
Comment (chelsea-lin) [inline bigframes/operations/aggregations.py:116]: same as below

File: bigframes/core/compile/scalar_op_compiler.py [hunk 6]
```diff
@@ -1070,8 +1070,16 @@ def floordiv_op(
     )
 
 
-def _is_float(x: ibis_types.Value):
-    return isinstance(x, (ibis_types.FloatingColumn, ibis_types.FloatingScalar))
+def _is_bignumeric(x: ibis_types.Value):
+    if not isinstance(x, ibis_types.DecimalValue):
+        return False
+    # Should be exactly 76 for bignumeric
+    return x.precision > 70
```
Comment (chelsea-lin) [inline bigframes/core/compile/scalar_op_compiler.py:1077]: Can you update it into `== 76` as you handle the numeric casting at line 1102 in this line?

**Review by chelsea-lin — APPROVED** [review]
Left few questions. LGTM overall.

File: bigframes/operations/aggregations.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/operations/aggregations.py:210]: yes, the test suite automatically validates all these derivations now

File: bigframes/core/compile/scalar_op_compiler.py (hunk 6 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/scalar_op_compiler.py:1077]: I care about the pre-cast type information, and I don't trust ibis to have the precision/scale numbers right.

File: bigframes/operations/aggregations.py (hunk 4 above)
Comment (TrevorBergeron) [reply bigframes/operations/aggregations.py:129]: done

File: bigframes/session/__init__.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/session/__init__.py:957]: done

**Pull request merged**